aws-api/src/main/java/com/amplifyframework/api/aws/SelectionSet.java

```diff
@@ -0,0 +1,185 @@
+
+    // This class should not be instantiated
+    private SelectionSet() {}
+
+    /**
+     * Returns selection set containing the fields of the provided model class, as well as nested models.
+     * @param modelClass model class
+     * @return selection set containing all of the fields of the provided model class
+     * @throws AmplifyException if a ModelSchema cannot be created from the provided model class.
+     */
+    public static Node fromModelClass(Class<? extends Model> modelClass, int depth) throws AmplifyException {
+        return new Node(null, getModelFields(modelClass, depth));
+    }
+
+    /**
+     * Expects a {@link Node} containing {@link Model} fields as nodes, and returns a new root node with two children:
+     *  - "items" with nodes being the children of the provided node.
+     *  - "nextToken"
+     *
+     * @param node a root node, with a value of null, and pagination fields
+     * @return
+     */
+    public static Node wrapPagination(Node node) {
+        return new Node(null, wrapPagination(node.nodes));
+    }
+
+    private static Set<Node> wrapPagination(Set<Node> nodes) {
+        Set<Node> paginatedSet = new HashSet<>();
+        paginatedSet.add(new Node(ITEMS_KEY, nodes));
+        paginatedSet.add(new Node(NEXT_TOKEN_KEY, null));
+        return paginatedSet;
+    }
+
+    @SuppressWarnings("unchecked") // Cast to Class<Model>
+    private static Set<Node> getModelFields(Class<? extends Model> clazz, int depth)
+            throws AmplifyException {
+        if (depth < 0) {
+            return new HashSet<>();
+        }
+
+        Set<Node> result = new HashSet<>();
+        ModelSchema schema = ModelSchema.fromModelClass(clazz);
+        for (Field field : FieldFinder.findFieldsIn(clazz)) {
+            String fieldName = field.getName();
+            if (schema.getAssociations().containsKey(fieldName)) {
+                if (List.class.isAssignableFrom(field.getType())) {
+                    if (depth >= 1) {
+                        ParameterizedType listType = (ParameterizedType) field.getGenericType();
+                        Class<Model> listTypeClass = (Class<Model>) listType.getActualTypeArguments()[0];
+                        Set<Node> fields = wrapPagination(getModelFields(listTypeClass, depth - 1));
+                        result.add(new Node(fieldName, fields));
+                    }
+                } else if (depth >= 1) {
+                    Set<Node> fields = getModelFields((Class<Model>) field.getType(), depth - 1);
+                    result.add(new Node(fieldName, fields));
+                }
+            } else {
+                result.add(new Node(fieldName, null));
+            }
+            for (AuthRule authRule : schema.getAuthRules()) {
+                if (AuthStrategy.OWNER.equals(authRule.getAuthStrategy())) {
+                    result.add(new Node(authRule.getOwnerFieldOrDefault(), null));
+                    break;
+                }
```

Comment:
on iOS, when generating the selection set, it's a bit more generic and only adds the fields from the Model schema to the selection set. We don't actually add it in based on the auth rule's owner field. The reason why we do not add it in is because it's up to the developer to decide whether that data is returned in their Model

For example
1. Developer has annotated the Model with Auth directive with "allow: owner" and by default it sets ownerField to "owner". They do not add it to their model like
```
type Model @model @auth { allow: owner } {
   id: ID!
   content: String
   // missing owner
}
```

Model only contains id and content, so if the selection set had the owner field added, it would just be dropped when deserializing
DataStore sync to cloud does custom deserialization to separate out into Model and MutationSyncMetadata (containinig version and deleted). Even if ownerField is added to the selection set, and the value is returned in the response, it would just be ignored

2. Developer sets it on their model
```
type Model @model @auth { allow: owner } {
   id: ID!
   content: String
   owner: String
}
```
Codegen generates Model with id, content, and owner. SelectionSet contains all three fields because it's generated from the Model's fields. Does not care about the auth rules

3. Developer sets a custom ownerField, but not on their model
```
type Model @model @auth { allow: owner, ownerField: "customOwner" } {
   id: ID!
   content: String
   // missing customOwner
}
```
I don't think a developer would ever do this since they are trying to reference a custom owner field in the Model but have not specified it in their model. If they want a custom owner field, they should specify it like in scenario 4 below

4. Developer sets a custom ownerField and adds it to the model
```
type Model @model @auth { allow: owner, ownerField: "customOwner" } {
   id: ID!
   content: String
   customOwner: String
}
```
Codegen generates Model with id, content, and customOwner. SelectionSet contains all three fields because it's generated from the Model's fields. Does not care about the auth rules.

**owner field on the Model should be optional**
when you specify the owner field on the Model, whether it's "owner" or "customOwner", it has to be optional since it's automatically managed (API intercept the request with the token). Making it required means the caller needs to add it in the request somehow.


For scenario 2 and 4, does that mean this code is adding duplicate owner fields to the selection set? I think this code can just be removed if it the selection set is already being constructed by model fields.

Reply:
this may not be relevant on Android depending on how the graphql variables are generated from the model instance, but on iOS, for scenario 2 and 4, we would generate the graphQL variables based on the model instance. Since the model instance contains the owner field, requests with the input with null owner failed.

Model instance created like `let model = Model(id = "111", owner = nil)` would generate a graphql variables like
```
{
   "input": {
      "id": "111",
      "owner": null
   }
}
```

So, we actually [filter it out](https://github.com/aws-amplify/amplify-ios/blob/eb5b7ff98ae9ac090dc8c71e374b8f3bcb0ee472/AmplifyPlugins/Core/AWSPluginsCore/Model/Decorator/ModelDecorator.swift#L31)

Reply:
> For scenario 2 and 4, does that mean this code is adding duplicate owner fields to the selection set?

Good question - no, it doesn't allow duplicates because the data structure is a `Set`.

That said, thank you for that explanation.  I agree we should just allow the developer to decide whether to include the `owner` field in the selection set, so I've removed the logic I had for adding it.

> So, we actually [filter the owner out](https://github.com/aws-amplify/amplify-ios/blob/eb5b7ff98ae9ac090dc8c71e374b8f3bcb0ee472/AmplifyPlugins/Core/AWSPluginsCore/Model/Decorator/ModelDecorator.swift#L31).

Thanks for sharing this.  I didn't have this before, but have added similar logic in this PR to filter out the owner field, if null, for mutations.